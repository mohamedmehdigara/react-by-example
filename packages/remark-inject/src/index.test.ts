import { promises } from 'fs'
import { Emphasis, Image, Text } from 'mdast'
import { extname } from 'path'
import remark from 'remark'
import remarkMdx from 'remark-mdx'
import { Processor } from 'unified'
import remarkInject from '.'

const { readFile } = promises

const fixtures = {
  simple: `\
# Hello world!

## Heading _level_ 2

![](@app/example-packages/example-packages/hello-world/src/index.tsx)

hello link _injector_

![](@app/example-packages/example-packages/static-clock/public/index.html)
`,
}

describe('remarkInject', () => {
  let processor: Processor

  beforeEach(() => {
    processor = remark()
      .use(remarkMdx)
      .use(remarkInject(), {
        visitor({ node, inject }) {
          const { children, type } = node as Emphasis

          if (type !== 'emphasis') return inject.nothing()

          const [{ value }] = children as Text[]
          return inject.link(value, `https://example.com/${value}`)
        },
      })
      .use(remarkInject(), {
        async visitor({ node, inject, create }) {
          const { type, url } = node as Image

          if (type !== 'image') return inject.nothing()
          if (typeof url !== 'string') return inject.nothing()

          const path = require.resolve(url)
          const lang = extname(path).slice(1)
          const code = await readFile(path, 'utf8')

          const result = [...inject.code(code, lang)]

          if (lang === 'tsx') {
            const identifier = create.defaultImport(url)
            result.push(...inject.jsx(identifier, { some: 'props' }))
          }

          return result
        },
      })
  })

  it.each(Object.entries(fixtures))(
    'should inject nodes "%s"',
    async (_name, markdown) => {
      const result = await processor.process(markdown)
      expect(result.toString()).toMatchSnapshot()
    },
  )
})
