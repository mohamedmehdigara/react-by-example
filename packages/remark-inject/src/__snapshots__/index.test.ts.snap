// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remarkInject should inject nodes "simple" 1`] = `
"import APP_EXAMPLE_PACKAGES_EXAMPLE_PACKAGES_HELLO_WORLD_SRC_INDEX_TSX_DEFAULT from '@app/example-packages/example-packages/hello-world/src/index.tsx'

# Hello world!

## Heading [level](https://example.com/level)_level_ 2

<!--@app/example-packages/example-packages/hello-world/src/index.tsx-->\`\`\`tsx
import React from 'react'
import ReactDOM from 'react-dom'

ReactDOM.render(<h1>Hello world!</h1>, document.getElementById('root'))

\`\`\`<>{APP_EXAMPLE_PACKAGES_EXAMPLE_PACKAGES_HELLO_WORLD_SRC_INDEX_TSX_DEFAULT}</>![](@app/example-packages/example-packages/hello-world/src/index.tsx)

hello link [injector](https://example.com/injector)_injector_

<!--@app/example-packages/example-packages/static-clock/public/index.html-->\`\`\`html
<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"utf-8\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />
    <title>React App</title>
  </head>
  <body>
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <div id=\\"root\\"></div>
  </body>
</html>

\`\`\`![](@app/example-packages/example-packages/static-clock/public/index.html)
"
`;
